e3sm_diags/derivations/formulas.py

```diff
@@ -474,3 +493,382 @@ def _replace_inf_with_nan(var: xr.DataArray) -> xr.DataArray:
+
+    See Ghan 2013 for derivation of ERF decomposition: https://doi.org/10.5194/acp-13-9971-2013
+    """
+    with xr.set_options(keep_attrs=True):
+        var = (fsnt - flnt) - (fsnt_d1 - flnt_d1)
+
+    var.attrs["units"] = "W/m2"
+    var.attrs["long_name"] = "ERFari: direct effect"
+
+    return var
+
+
+def erf_aci(
+    fsnt_d1: xr.DataArray,
+    flnt_d1: xr.DataArray,
+    fsntc_d1: xr.DataArray,
+    flntc_d1: xr.DataArray,
+) -> xr.DataArray:
+    """
+    Calculate aerosol--cloud interactions (ACI) part of effectie radiative forcing (ERF)
+
+    Parameters
+    ----------
+    fsnt_d1 : xr.DataArray
+        Downward shortwave radiation toa without aerosols.
+    flnt_d1 : xr.DataArray
+        Upward longwave radiation toa without aerosols.
+    fsntc_d1 : xr.DataArray
+        fsnt_d1 without clouds.
+    flntc_d1 : xr.DataArray
+        flnt_d1 without clouds.
+
+    Returns
+    -------
+    xr.DataArray
+        ERFaci (aka, indirect effect) in W/m2.
+
+    Notes
+    -----
+    See Ghan 2013 for derivation of ERF decomposition: https://doi.org/10.5194/acp-13-9971-2013
+    """
+    with xr.set_options(keep_attrs=True):
+        var = (fsnt_d1 - flnt_d1) - (fsntc_d1 - flntc_d1)
+
+    var.attrs["units"] = "W/m2"
+    var.attrs["long_name"] = "ERFaci: indirect effect"
+
+    return var
+
+
+def erf_res(fsntc_d1: xr.DataArray, flntc_d1: xr.DataArray) -> xr.DataArray:
+    """
+    Calculate the residual effect (RES) part of effective radiative forcin g.
+
+    Parameters
+    ----------
+    fsntc_d1 : xr.DataArray
+        Downward solar radiation at the top of the atmosphere with neither
+        clouds nor aerosols.
+    flntc_d1 : xr.DataArray
+        Upward longwave radiation at the top of the atmosphere with neither
+        clouds nor aerosols.
+
+    Returns
+    -------
+    xr.DataArray
+        ERFres (aka, surface effect) in W/m2.
+
+    Notes
+    -----
+    See Ghan 2013 for derivation of ERF decomposition: https://doi.org/10.5194/acp-13-9971-2013
+    """
+    with xr.set_options(keep_attrs=True):
+        var = fsntc_d1 - flntc_d1
+
+    var.attrs["units"] = "W/m2"
+    var.attrs["long_name"] = "ERFres: residual effect"
+
+    return var
+
+
+def sum_vars(vars: List[xr.DataArray]) -> xr.DataArray:
+    """Sum DataArrays using Python's `.sum()` and perserve attrs.
+
+    Pythons sum iterates over the iterable (the list of DataArrays) and
+    adds all elements, which is different from NumPy which performs a sum
+    reduction over an axis/axes. This function ensures the DataArray attributes
+    are perserved by invoking the `.sum()` call within the context of
+    `xr.set_options()`.
+
+    Parameters
+    ----------
+    vars : List[xr.DataArray]
+        A list of variables.
+
+    Returns
+    -------
+    xr.DataArray
+        The sum of the variables
+    """
+    with xr.set_options(keep_attrs=True):
+        result: xr.DataArray = sum(vars)  # type: ignore
+
+    return result
```

Comment:
The new `sum_vars()` function replaces the original `lambda(*x): sum(x)` and `lambda(*x): molec_convert_units(sum(x), 12)` functions defined in the derived variables dictionary.

It places the call to `sum(vars)` inside `with xr.set_options(keep_attrs=True)` to maintain attributes, including `"units"` which is required for `molec_convert_units()`.